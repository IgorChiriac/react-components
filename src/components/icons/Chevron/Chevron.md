This component is a wrapper around the svg code for Chevron's icon.

Allowed directions are 'down' | 'up' | 'left' | 'right' | number | string

```jsx
<Chevron color="blue" direction="down"/>
<Chevron color="blue" direction="up"/>
<Chevron color="blue" direction="left"/>
<Chevron color="blue" direction="right"/>
<Chevron color="blue" direction={45}/>
<Chevron color="blue" direction="225"/>
```
