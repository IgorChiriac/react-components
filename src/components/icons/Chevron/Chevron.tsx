import React from 'react';
import styled from 'styled-components';
import * as colors from '../../../constants/colors';

export type ChevronDirection = 'up' | 'down' | 'left' | 'right' | number | string;

export interface IChevronProps extends React.SVGProps<SVGSVGElement> {
  /**
   * Targets the svg's `fill` attribute
   * @default colors.extended.blue300
   */
  color?: string;

  /**
   * Chevron arrow direction
   * @default down
   */
  direction?: ChevronDirection;
}

const getRotation = (direction: ChevronDirection = 'down'): string => {
  const directionMap = {
    down: '0',
    left: '90',
    right: '-90',
    up: '180',
  };
  return directionMap[direction] || direction;
};

const StyledChevron = styled.svg<IChevronProps>`
  transition: transform 0.2s;
  transform: rotate(${props => getRotation(props.direction)}deg);
`;

// ref is excluded from ...rest due to a type conflict; TODO: check if styled-components v4 fixes it.
const Chevron = ({
  ref,
  color = colors.extended.blue300,
  height = '24px',
  width = '24px',
  direction = 'down',
  ...rest
}: IChevronProps) => (
  <StyledChevron
    width={width}
    height={height}
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 416 416"
    direction={direction}
    {...rest}
  >
    <path
      d="M214.8 287c2.2-.9 4.3-2.3 6.1-4.1l124-124c7.2-7.1 7.2-18.7 0-25.9-7.1-7.2-18.7-7.2-25.9 0L207.9 244.1l-111-111c-7.1-7.2-18.7-7.2-25.9 0-7.2 7.1-7.2 18.7 0 25.9l124 124c5.4 5.3 13.2 6.7 19.8 4z"
      fill={color}
    />
  </StyledChevron>
);

export default Chevron;
