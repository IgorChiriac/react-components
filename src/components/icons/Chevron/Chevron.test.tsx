import React from 'react';
import { render } from 'react-testing-library';
import Chevron from './Chevron';

describe('<Chevron />', () => {
  const directions = ['down', 'up', 'left', 'right', 45, '225'];
  directions.forEach(direction => {
    it(`renders the component pointing ${direction}`, () => {
      const { container } = render(<Chevron color="black" direction={direction} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
