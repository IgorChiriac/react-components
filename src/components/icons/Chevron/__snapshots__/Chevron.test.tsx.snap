// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chevron /> renders the component pointing 45 1`] = `
.c0 {
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

<svg
  class="c0"
  direction="45"
  height="24px"
  viewBox="0 0 416 416"
  width="24px"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M214.8 287c2.2-.9 4.3-2.3 6.1-4.1l124-124c7.2-7.1 7.2-18.7 0-25.9-7.1-7.2-18.7-7.2-25.9 0L207.9 244.1l-111-111c-7.1-7.2-18.7-7.2-25.9 0-7.2 7.1-7.2 18.7 0 25.9l124 124c5.4 5.3 13.2 6.7 19.8 4z"
    fill="black"
  />
</svg>
`;

exports[`<Chevron /> renders the component pointing 225 1`] = `
.c0 {
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: rotate(225deg);
  -ms-transform: rotate(225deg);
  transform: rotate(225deg);
}

<svg
  class="c0"
  direction="225"
  height="24px"
  viewBox="0 0 416 416"
  width="24px"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M214.8 287c2.2-.9 4.3-2.3 6.1-4.1l124-124c7.2-7.1 7.2-18.7 0-25.9-7.1-7.2-18.7-7.2-25.9 0L207.9 244.1l-111-111c-7.1-7.2-18.7-7.2-25.9 0-7.2 7.1-7.2 18.7 0 25.9l124 124c5.4 5.3 13.2 6.7 19.8 4z"
    fill="black"
  />
</svg>
`;

exports[`<Chevron /> renders the component pointing down 1`] = `
.c0 {
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

<svg
  class="c0"
  direction="down"
  height="24px"
  viewBox="0 0 416 416"
  width="24px"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M214.8 287c2.2-.9 4.3-2.3 6.1-4.1l124-124c7.2-7.1 7.2-18.7 0-25.9-7.1-7.2-18.7-7.2-25.9 0L207.9 244.1l-111-111c-7.1-7.2-18.7-7.2-25.9 0-7.2 7.1-7.2 18.7 0 25.9l124 124c5.4 5.3 13.2 6.7 19.8 4z"
    fill="black"
  />
</svg>
`;

exports[`<Chevron /> renders the component pointing left 1`] = `
.c0 {
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

<svg
  class="c0"
  direction="left"
  height="24px"
  viewBox="0 0 416 416"
  width="24px"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M214.8 287c2.2-.9 4.3-2.3 6.1-4.1l124-124c7.2-7.1 7.2-18.7 0-25.9-7.1-7.2-18.7-7.2-25.9 0L207.9 244.1l-111-111c-7.1-7.2-18.7-7.2-25.9 0-7.2 7.1-7.2 18.7 0 25.9l124 124c5.4 5.3 13.2 6.7 19.8 4z"
    fill="black"
  />
</svg>
`;

exports[`<Chevron /> renders the component pointing right 1`] = `
.c0 {
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

<svg
  class="c0"
  direction="right"
  height="24px"
  viewBox="0 0 416 416"
  width="24px"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M214.8 287c2.2-.9 4.3-2.3 6.1-4.1l124-124c7.2-7.1 7.2-18.7 0-25.9-7.1-7.2-18.7-7.2-25.9 0L207.9 244.1l-111-111c-7.1-7.2-18.7-7.2-25.9 0-7.2 7.1-7.2 18.7 0 25.9l124 124c5.4 5.3 13.2 6.7 19.8 4z"
    fill="black"
  />
</svg>
`;

exports[`<Chevron /> renders the component pointing up 1`] = `
.c0 {
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

<svg
  class="c0"
  direction="up"
  height="24px"
  viewBox="0 0 416 416"
  width="24px"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M214.8 287c2.2-.9 4.3-2.3 6.1-4.1l124-124c7.2-7.1 7.2-18.7 0-25.9-7.1-7.2-18.7-7.2-25.9 0L207.9 244.1l-111-111c-7.1-7.2-18.7-7.2-25.9 0-7.2 7.1-7.2 18.7 0 25.9l124 124c5.4 5.3 13.2 6.7 19.8 4z"
    fill="black"
  />
</svg>
`;
